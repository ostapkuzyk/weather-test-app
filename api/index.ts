// https://home.openweathermap.org/
// email:weather.test@mailto.plus
// pass:SXXm+5$y!3wfN&F
// apiKey:15ccbcc5f047f14b04d1ada34df4c710
// -----------------------
// https://www.weatherapi.com/
// email:weather.test@mailto.plus
// pass:@4zNuJ%VCUDj$sK
// apiKey:2177b147ec464aeb83891514243112 (for 14 Jan)
import axios from 'axios';

interface WeatherServiceConfig {
  apiKey: string;
  baseUrl: string;
}

console.log('process.env.WEATHERAPI_API_KEY: ', process.env.EXPO_PUBLIC_OPENWEATHER_API_KEY);
export const weatherServiceConfigs = {
  openWeatherMap: {
    apiKey: process.env.EXPO_PUBLIC_OPENWEATHER_API_KEY || '',
    baseUrl: 'https://api.openweathermap.org',
  } as WeatherServiceConfig,
  weatherApi: {
    apiKey: process.env.EXPO_PUBLIC_WEATHERAPI_API_KEY || '',
    baseUrl: 'https://api.weatherapi.com/v1',
  } as WeatherServiceConfig,
};

export const createAxiosInstance = (service: keyof typeof weatherServiceConfigs) => {
  const { apiKey, baseUrl } = weatherServiceConfigs[service];

  const axiosInstance = axios.create({
    baseURL: baseUrl,
    timeout: 10000,
    headers: {
      'Content-Type': 'application/json',
    },
  });

  axiosInstance.interceptors.request.use(
    (config) => {
      if (service === 'openWeatherMap') {
        config.params = config.params || {};
        config.params.appid = apiKey;
      } else if (service === 'weatherApi') {
        config.params = config.params || {};
        config.params.key = apiKey;
      }

      return config;
    },
    (error) => {
      return Promise.reject(error);
    }
  );

  axiosInstance.interceptors.response.use(
    (response) => response.data,
    (error) => {
      if (error.response) {
        console.error('Response error:', error.response.data);
      } else if (error.request) {
        console.error('Request error:', error.request);
      } else {
        console.error('Error message:', error.message);
      }
      return Promise.reject(error);
    }
  );

  return axiosInstance;
};
