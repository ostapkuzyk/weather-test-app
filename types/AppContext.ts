export type TAddress = {
  name: string;
  lat: number;
  lon: number;
};

export interface AppContextType {
  address: TAddress;
  handleChangeAddress: (value: TAddress) => void;
  weatherService: WeatherService;
  setWeatherService: (service: WeatherService) => void;
}

export enum WeatherService {
  OpenWeatherMap = 'openWeatherMap',
  WeatherApi = 'weatherApi',
}
