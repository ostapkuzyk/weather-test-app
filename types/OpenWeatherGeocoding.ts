interface LocalNames {
  [key: string]: string;
}

interface Location {
  country: string;
  lat: number;
  lon: number;
  local_names?: LocalNames;
  name: string;
  state?: string;
}
