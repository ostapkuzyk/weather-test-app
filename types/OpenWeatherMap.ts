export interface WeatherResponse {
  lat: number;
  lon: number;
  timezone: string;
  timezone_offset: number;
  current: CurrentWeather;
  daily: DailyWeather[];
}

export interface CurrentWeather {
  dt: number;
  sunrise: number;
  sunset: number;
  temp: number;
  feels_like: number;
  pressure: number;
  humidity: number;
  dew_point: number;
  uvi: number;
  clouds: number;
  visibility: number;
  wind_speed: number;
  wind_deg: number;
  wind_gust: number;
  weather: WeatherDescription[];
}

export interface DailyWeather {
  dt: number;
  sunrise: number;
  sunset: number;
  moonrise: number;
  moonset: number;
  moon_phase: number;
  summary: string;
  temp: Temperature;
  feels_like: FeelsLike;
  pressure: number;
  humidity: number;
  dew_point: number;
  wind_speed: number;
  wind_deg: number;
  wind_gust: number;
  weather: WeatherDescription[];
  clouds: number;
  pop: number;
  uvi: number;
  snow?: number;
}

export interface WeatherDescription {
  id: number;
  main: string;
  description: string;
  icon: string;
}

export interface Temperature {
  day: number;
  min: number;
  max: number;
  night: number;
  eve: number;
  morn: number;
}

export interface FeelsLike {
  day: number;
  night: number;
  eve: number;
  morn: number;
}
