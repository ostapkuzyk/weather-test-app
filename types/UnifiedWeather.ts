export interface UnifiedWeatherResponse {
  current: {
    temperature: number;
    feelsLike: number;
    humidity: number;
    windSpeed: number;
    condition: string;
    pressure: number;
    icon: string;
  };
  forecast: UnifiedWeatherForecast[];
}

export interface UnifiedWeatherForecast {
  date: string;
  maxTemp: number;
  minTemp: number;
  condition: string;
  icon: string;
  precipitation: number;
  windSpeed: number;
}
