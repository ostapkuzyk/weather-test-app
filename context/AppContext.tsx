import React, { createContext, ReactNode, useContext, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { STORAGE_ADDRESS_KEY, STORAGE_WEATHER_SERVICE_KEY } from '@/constants/asyncStorageKeys';
import { AppContextType, TAddress, WeatherService } from '@/types/AppContext';
import { initAsyncStorage } from '@/constants/mockData';

const AppContext = createContext<AppContextType | undefined>(undefined);

export const AppProvider: React.FC<{
  children: ReactNode;
}> = ({ children }) => {
  const [address, setAddress] = useState<TAddress>(initAsyncStorage);
  const [weatherService, setWeatherService] = useState<WeatherService>(WeatherService.WeatherApi);

  useEffect(() => {
    const loadInitialData = async () => {
      try {
        // Load address
        const addressJson = await AsyncStorage.getItem(STORAGE_ADDRESS_KEY);
        const parsedAddress: TAddress | null = addressJson != null ? JSON.parse(addressJson) : null;
        if (parsedAddress) {
          setAddress(parsedAddress);
        }

        // Load weather service
        const service = await AsyncStorage.getItem(STORAGE_WEATHER_SERVICE_KEY);
        const parsedService: WeatherService | null = service != null ? JSON.parse(service) : null;

        if (
          parsedService &&
          Object.values(WeatherService).includes(parsedService as WeatherService)
        ) {
          setWeatherService(parsedService as WeatherService);
        }
      } catch (error) {
        console.error('Error loading data from AsyncStorage:', error);
      }
    };

    loadInitialData();
  }, []);

  const handleChangeAddress = async (newAddress: TAddress) => {
    try {
      const jsonValue = JSON.stringify(newAddress);
      await AsyncStorage.setItem(STORAGE_ADDRESS_KEY, jsonValue);
      setAddress(newAddress);
    } catch (error) {
      console.error('Failed to save address to AsyncStorage:', error);
    }
  };

  const handleSetWeatherService = async (service: WeatherService) => {
    try {
      const jsonValue = JSON.stringify(service);
      await AsyncStorage.setItem(STORAGE_WEATHER_SERVICE_KEY, jsonValue);
      setWeatherService(service);
    } catch (error) {
      console.error('Failed to save weather service to AsyncStorage:', error);
    }
  };

  const contextValues: AppContextType = {
    address,
    handleChangeAddress,
    weatherService,
    setWeatherService: handleSetWeatherService,
  };

  return <AppContext.Provider value={contextValues}>{children}</AppContext.Provider>;
};

export const useAppContext = () => {
  const context = useContext(AppContext);
  if (!context) {
    throw new Error('useAppContext must be used within an AppProvider');
  }
  return context;
};