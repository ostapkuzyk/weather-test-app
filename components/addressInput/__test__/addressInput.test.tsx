import { WeatherService } from '@/types/AppContext';
import AddressInput from '../index';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { useAxiosInstance } from '@/api';

const mockService = WeatherService.OpenWeatherMap;

jest.mock('@/context/AppContext', () => ({
  useAppContext: jest.fn().mockReturnValue({
    weatherService: mockService,
    address: {
      name: 'name',
      lat: 1,
      lon: 1
    }
  })
}));

jest.mock('@/api', () => ({
  useAxiosInstance: jest.fn().mockReturnValue({
    fetchGeoCodingData: jest.fn()
  })
}));

jest.mock('expo-router', () => ({
  useRouter: jest.fn().mockReturnValue({
    back: jest.fn()
  })
}));

describe('addressInput', () => {
  beforeEach(() => {
    jest.useRealTimers();
  })
  it('renders without crashing', () => {
    const { toJSON } = render(<AddressInput
      address={''}
      setAddress={jest.fn()}
      onSave={jest.fn()} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('fetch data and selects one address', async () => {
    jest.useFakeTimers();
    const mockAddress = 'mock initial'
    const mockSetAddress = jest.fn()
    const mockOnSave = jest.fn();

    const mockInputValue = 'new value'

    const mockLocation: Location = {
      country: 'country',
      lat: 1,
      lon: 1,
      name: mockInputValue,
      state: 'state'
    };
    const mockResponseData = {
      data: [mockLocation]
    }

    const axiosInstance = useAxiosInstance(mockService);
    axiosInstance.fetchGeoCodingData.mockResolvedValueOnce(mockResponseData);

    const component = <AddressInput
      address={mockAddress}
      setAddress={mockSetAddress}
      onSave={mockOnSave} />
    const { findByTestId } = render(component);

    const inputField = await findByTestId('input');
    expect(inputField).toBeTruthy();

    await waitFor(() => {
      fireEvent.changeText(inputField, mockInputValue);
    });

    jest.advanceTimersByTime(1000);
    
    const resultItem = await findByTestId('resultItem');

    fireEvent.press(resultItem);

    expect(mockOnSave).toHaveBeenCalled();
    expect(mockSetAddress).toHaveBeenCalledWith(mockInputValue);
  });
});