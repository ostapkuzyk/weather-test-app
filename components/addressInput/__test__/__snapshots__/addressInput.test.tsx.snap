// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addressInput renders without crashing 1`] = `
<View>
  <Text
    style={
      {
        "color": "#fff",
        "fontSize": 16,
        "marginBottom": 10,
      }
    }
  >
    Enter your address
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Enter address"
    placeholderTextColor="#888"
    style={
      [
        {
          "backgroundColor": "#0B3B8C",
          "borderColor": "#0B3B8C",
          "borderRadius": 12,
          "borderWidth": 1,
          "color": "#fff",
          "fontSize": 16,
          "padding": 12,
        },
      ]
    }
    testID="input"
    value=""
  />
  <RCTScrollView
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View />
  </RCTScrollView>
</View>
`;
